File: epfl-hc/hc2012/graders/CryptoLib3.java
import java.math.BigInteger;
import java.util.Scanner;
import java.util.Random;

public class CryptoLib3 {

  private static Crypto3 crypto3;
  private static Scanner scanner;
  private static int x_n = 773137898;

  public static void main(String args[]) {
    crypto3 = new Crypto3();
    scanner = new Scanner(System.in);

    testDecrypt();

    System.out.println();
    System.out.println("OK - a03cf946a04f9ae2fd30c3e7d20a7009");
  }

  private static int my_rand()
  {
    final int A = 1103515245;
    final int C = 12345;
    x_n = A*x_n + C;

    int res = (x_n & 0x3FFFFFFF) / (65536);
    return res;
  }

  private static void unitTest() {
    // Note: in the real test, p will change all the time
    BigInteger p = BigInteger.valueOf(1000003);
    Random random = new Random();
    int correct = 0;
    int ntc = 100;
    for(int i=0;i<ntc;++i) {
      // Note: in the real test: g will be distributed uniformly at random
      // in its domain
      BigInteger g = (new BigInteger(19, random)).add(BigInteger.valueOf(2));
      BigInteger x = new BigInteger(19, random);
      BigInteger r = new BigInteger(19, random);
      BigInteger m1 = new BigInteger(19, random);
      BigInteger m2 = new BigInteger(19, random);
      BigInteger y = g.modPow(x, p);

      PublicKey pk = new PublicKey();
      pk.p = p;
      pk.g = g;
      pk.y = y;

      Ciphertext ct = encrypt(m1, pk, r);

      BigInteger m;
      if (Math.random()>0.5) {
        m = crypto3.decrypt(ct, pk, m1, m2);
      } else {
        m = crypto3.decrypt(ct, pk, m2, m1);
      }

      if(m.equals(m1)) {
        correct++;
      } else {
      }
    }
    System.out.println("You were correct " + correct*100./ntc + "% of the time.");
    System.out.println("(You should aim for 75% correct.)");
  }

  static Ciphertext encrypt(BigInteger plaintext, PublicKey key, BigInteger r) {
    Ciphertext ct = new Ciphertext();
    // E = m * y^r (mod p)
    ct.E = plaintext.multiply(key.y.modPow(r, key.p)).mod(key.p);
    // F = g^(-r)
    ct.F = key.g.modPow(r.negate(), key.p);
    return ct;
  }

  private static void testDecrypt()
  {
    /*
      We want that an algorithm with winning probability 24/32 (=75%)
        wins the game with probability 99.9%
      And that an algorithm with winning probability 23/32 (=71.8%)
        loses the game with probability 99.9%.

      Octave code:
      octave:43> N=7646;X=1:N;[i,iw]=max((binocdf(X, N, 24/32)<=0.001)+(binocdf(X, N, 23/32)>0.999))
      i =  2
      iw =  5616

      octave:45> [binocdf(iw,N,23/32) 1-binocdf(iw,N,24/32)]
      ans =

         0.99901   0.99903

      So with N = 7646 and requiring 5616 tests correct these are the winning probabilities:
      algo 24/32:   passes with probability 0.99903
      algo 23/32:   does not pass with probability 0.99901
    */
    int correct = 0;
    int mincorrect = 5616;
    /*
      Input format is:
      - Number of testcases
      - For each testcase:   E  F  p  g  y  x  m  m2
    */
    int ntc = scanner.nextInt();
    if(ntc != 7646) {
        System.err.println("ERROR: expected exactly 7646 testcases");
        System.exit(1);
    }
    for(int i=0;i<ntc;++i) {
      Ciphertext ct = new Ciphertext();
      ct.E = new BigInteger(scanner.next());
      ct.F = new BigInteger(scanner.next());
      PublicKey pk = new PublicKey();
      pk.p = new BigInteger(scanner.next());
      pk.g = new BigInteger(scanner.next());
      pk.y = new BigInteger(scanner.next());
      BigInteger x = new BigInteger(scanner.next());
      if(! pk.g.modPow(x, pk.p).equals(pk.y)) {
        System.err.println("ERROR: g^x != y  (mod p)");
        System.exit(1);
      }
      BigInteger m1 = new BigInteger(scanner.next());
      BigInteger m2 = new BigInteger(scanner.next());

      BigInteger m;
      if (my_rand()%2 == 0) {
        m = crypto3.decrypt(ct, pk, m1, m2);
      } else {
        m = crypto3.decrypt(ct, pk, m2, m1);
      }

      System.out.print(".");
      if(m.equals(m1)) {
        correct++;
      } else {
      }
    }

    if(correct < mincorrect) {
      System.out.println();
      System.out.println("Not enough correct answers: " + correct + " expected at least: " + mincorrect);
      System.out.println("Peformance: " + correct*100./ntc + "% expected performance: " + mincorrect*100./ntc + "%");
      System.exit(0);
    }
  }

}

class Ciphertext {
  public BigInteger E;
  public BigInteger F;
}

class PublicKey {
  public BigInteger p;
  public BigInteger g;
  public BigInteger y;
}

File: epfl-hc/hc2012/graders/cryptoLib3.cpp
#include <stdio.h>
#include <stdlib.h>
#include <gmp.h>
#include <time.h>

#include "cryptoLib3.hpp"

static int x_n = 773137898;
static int my_rand()
{
  const int A = 1103515245;
  const int C = 12345;
  int res;
  x_n = A*x_n + C;

  res = (x_n & 0x3FFFFFFF) / (65536);
  return res;
}
Ciphertext encrypt(Plaintext plaintext, PublicKey key, mpz_t r)
{
  Ciphertext ct;
  mpz_init(ct.E);
  
  /* E = m * y^r (mod p) */
  mpz_powm(ct.E, key.y, r, key.p);
  mpz_mul(ct.E, ct.E, plaintext.m);
  mpz_mod(ct.E, ct.E, key.p);

  mpz_init(ct.F);
  /* F = g^(-r) */
  mpz_neg(ct.F, r);
  mpz_powm(ct.F, key.g, ct.F, key.p);
  
  return ct;
}

#if 0
static void unitTest() {
    gmp_randstate_t rs;
    mpz_t x, r;
    PublicKey pk;
    mpz_t pminus1, pminus2;
    Plaintext m1, m2;
    int correct = 0;
    int ntc = 100;
    int i;

    /* Note: in the real test, p will change all the time */
    mpz_init_set_si(pk.p, 1000003);
    mpz_init_set_si(pminus1, 1000003-1);
    mpz_init_set_si(pminus2, 1000003-2);
    gmp_randinit_default(rs);
    mpz_init(pk.g);
    mpz_init(pk.y);
    mpz_init(x);
    mpz_init(r);
    mpz_init(m1.m);
    mpz_init(m2.m);
    srand(time(0));
    for(i=0;i<ntc;++i) {
      Plaintext m;
      Ciphertext ct;

      mpz_urandomm(pk.g, rs, pminus2);
      mpz_add_ui(pk.g, pk.g, 2);
      mpz_urandomm(x, rs, pminus1);
      mpz_urandomm(r, rs, pminus1);
      mpz_urandomm(m1.m, rs, pk.p);
      mpz_urandomm(m2.m, rs, pk.p);
      mpz_powm(pk.y, pk.g, x, pk.p);

      ct = encrypt(m1, pk, r);

      if (rand()%2==0) {
        m = decrypt(ct, pk, m1, m2);
      } else {
        m = decrypt(ct, pk, m2, m1);
      }

      if(mpz_cmp(m.m, m1.m)==0) {
        correct++;
      } else {
      }

      mpz_clear(ct.E);
      mpz_clear(ct.F);
    }
    mpz_clear(pk.p);
    mpz_clear(pminus1);
    mpz_clear(pminus2);
    mpz_clear(pk.g);
    mpz_clear(pk.y);
    mpz_clear(x);
    mpz_clear(r);
    mpz_clear(m1.m);
    mpz_clear(m2.m);
    gmp_randclear(rs);
    printf("You were correct %3.0f%% of the time.\n", correct*100./ntc);
    printf("(You should aim for 75%% correct.)\n");
}
#endif

static void testDecrypt() {
    int correct = 0;
    int mincorrect = 5616;
    /*
      Input format is:
      - Number of testcases
      - For each testcase:   E  F  p  g  y  x  m  m2
    */
    int ntc, i;
    if(1!=scanf("%d\n", &ntc)) {
      printf("Input failure.\n");
      exit(1);
    }
    for(i=0;i<ntc;++i) {
      Ciphertext ct;
      PublicKey pk;
      Plaintext got;
      mpz_t x;
      Plaintext m1, m2, mleft, mright;
      mpz_init(ct.E);
      mpz_init(ct.F);
      mpz_init(pk.p);
      mpz_init(pk.g);
      mpz_init(pk.y);
      mpz_init(x);
      mpz_init(m1.m);
      mpz_init(m2.m);
      mpz_init(mleft.m);
      mpz_init(mright.m);
      if(8!=gmp_scanf("%Zd %Zd %Zd %Zd %Zd %Zd %Zd %Zd\n", ct.E, ct.F, pk.p, pk.g, pk.y, x, m1.m, m2.m)) {
        printf("Input failure %d.\n", i);
        exit(1);
      }

      if (my_rand()%2 == 0) {
        mpz_set(mleft.m, m1.m);
        mpz_set(mright.m, m2.m);
      } else {
        mpz_set(mleft.m, m2.m);
        mpz_set(mright.m, m1.m);
      }
      got = decrypt(ct, pk, mleft, mright);

      if(mpz_cmp(got.m, m1.m)==0) {
        correct++;
      } else {
      }

      printf(".");
      mpz_clear(ct.E);
      mpz_clear(ct.F);
      mpz_clear(pk.p);
      mpz_clear(pk.g);
      mpz_clear(pk.y);
      mpz_clear(x);
      mpz_clear(m1.m);
      mpz_clear(m2.m);
      mpz_clear(mleft.m);
      mpz_clear(mright.m);
    }
    if(correct < mincorrect) {
      printf("\n");
      printf("Not enough correct answers: %d expected at least: %d\n", correct, mincorrect);
      printf("Peformance: %3.5f%% expected performance: %3.5f%%\n", correct*100./ntc, mincorrect*100./ntc);
      exit(0);
    }
}

int main(int argc, char** argv) {
  /*unitTest();*/
  testDecrypt();

  printf("\n");
  printf("OK - a03cf946a04f9ae2fd30c3e7d20a7009\n");
  return 0;
}
